### PR TITLE
XML variable substitution - Escape XML characters fix

### LINKED ISSUE #3495: AzureRmWebAppDeploy: Xml Variable Substitution does not work with connectionStrings
We are using the 3.x preview AzureRmWebAppDeploy task. All app settings are replaced as expected. However, connection strings are not substituted. Looking at **webdeployment-common/xmlvariablesubstitutionutility.ts**, it looks like it's only looking for attributes named **key** and **value**, whereas the connectionStrings section uses **name** and **connectionString**.

### PR DESCRIPTION
**Pull request by vincentdass**
#3495 .

### CODE REVIEW

File: Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts [hunk 1]
```diff
@@ -138,60 +141,60 @@ export async function substituteXmlVariables(configFile, tags, variableMap, para
...
-            tl.debug("Config file " + configFile + " updated.");
-        }
-    });
-    
+    var domContent = ( fileEncodeType[1]? '\uFEFF' : '' ) + ltxdomutility.getContentWithHeader(xmlDocument);
+    for(var replacableTokenValue in replacableTokenValues) {
+        tl.debug('Substituting original value in place of temp_name: ' + replacableTokenValue);
+        domContent = domContent.split(replacableTokenValue).join(replacableTokenValues[replacableTokenValue]);
+    }
+    tl.writeFile(configFile, domContent, fileEncodeType[0]);
```
Comment (Ajay-MS) [inline Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts]: Does tl.writeFile  prints file name in error message if it failed to write content back to file ?

Comment (vincentdass) [inline Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts]: yes, it internally calls '[fs.writeFileSync](https://github.com/Microsoft/vsts-task-lib/blob/master/node/task.ts#L502)' which throws errors.

File: Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts [hunk 2]
```diff
@@ -138,60 +141,60 @@ export async function substituteXmlVariables(configFile, tags, variableMap, para
...
-        if (error) {
-            throw new Error(tl.loc("Failedtowritetoconfigfilewitherror",configFile, error));
-        } else {
-            tl.debug("Config file " + configFile + " updated.");
-        }
-    });
-    
+    var domContent = ( fileEncodeType[1]? '\uFEFF' : '' ) + ltxdomutility.getContentWithHeader(xmlDocument);
+    for(var replacableTokenValue in replacableTokenValues) {
+        tl.debug('Substituting original value in place of temp_name: ' + replacableTokenValue);
```
Comment (Ajay-MS) [inline Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts]: As discussed check on complexity part of algorithm as there is much difference is complexity when substituting in string or in dom.

File: Tasks/Common/webdeployment-common/Tests/L1XmlVarSub.ts [hunk 3]
```diff
@@ -10,7 +10,7 @@ async function xmlVarSub() {
         'webpages:Version' : '1.1.7.3',
         'xdt:Transform' : 'DelAttributes',
         'xdt:Locator' : 'Match(tag)',
-        'DefaultConnection': 'connectionStringValue1',
+        'DefaultConnection': "Url=https://primary;Database=db1;ApiKey=11111111-1111-1111-1111-111111111111;Failover = {Url:'https://secondary', ApiKey:'11111111-1111-1111-1111-111111111111'}",
```
Comment (KrishnaAdityaB) [inline Tasks/Common/webdeployment-common/Tests/L1XmlVarSub.ts]: can you add '<', '>', '[[' chars to this string

File: Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts [hunk 4]
```diff
@@ -22,55 +22,56 @@ function getReplacableTokenFromTags(xmlNode, variableMap) {
...
     }
-
     var fileBuffer: Buffer = fs.readFileSync(parameterFilePath);
     var fileEncodeType = fileEncoding.detectFileEncoding(parameterFilePath, fileBuffer);
     var webConfigContent: string = fileBuffer.toString(fileEncodeType[0]);
     if(fileEncodeType[1]) {
         webConfigContent = webConfigContent.slice(1);
     }
     var xmlDocument;
+    const paramFileReplacableToken = 'PARAM_FILE_REPLACE_TOKEN';
```
Comment (KrishnaAdityaB) [inline Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts]: can you prefix this with VSTS_

File: Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts [hunk 5]
```diff
@@ -22,55 +22,56 @@ function getReplacableTokenFromTags(xmlNode, variableMap) {
...
             }
         }
     }
     catch(error) {
         tl.debug("Unable to parse parameter file : " + parameterFilePath + ' Error: ' + error);
         return;
     }
-
     var domContent = (fileEncodeType[1] ? '\uFEFF' : '') + ltxdomutility.getContentWithHeader(xmlDocument);
-
```
Comment (KrishnaAdityaB) [inline Tasks/Common/webdeployment-common/xmlvariablesubstitutionutility.ts]: why did you remove all these empty lines. this is a very big function and those separators were increasing the readability

**Pull request merged**